benches/vm/mod.rs

```diff
@@ -10,12 +14,14 @@ pub fn load_vm_hello_world(c: &mut Criterion) {
 		memory_size: Byte::from_u64(1024 * 100_000).try_into().unwrap(),
 		..Default::default()
 	};
-	let mut vm = Uhyve::new(path, params).expect("Unable to create VM");
+
+	#[cfg(target_os = "linux")]
+	let mut vm = UhyveVm::<KvmCpu>::new(path, params).expect("Unable to create VM");
+	#[cfg(target_os = "macos")]
+	let mut vm = UhyveVm::<XhyveCpu>::new(path, params).expect("Unable to create VM");
```

Comment:
Interestingly, this is not necessarily always required. See `src/bin/uhyve.rs`. As there is only one implementation for `VirtualCpuType` present for each architecture, Rust apparently picks that one.

I'm not quite sure if the verbosity is really a problem. As for our use cases, the virtualization backend does matter, it might make sense if the user should be forced to think about it. I'm still undecided, though.

Reply:
The benchmarks you picked here are probably a pretty good example of where it should be kept explicit. If one day the default changes, you wouldn't want a surprising change in the benchmark performance just because it uses a different backend.

Reply:
The reason explicit types are only required here is that `run` is only implemented for the concrete type:

```rust
impl<VCpuType: VirtualCPU> UhyveVm<VCpuType> {
	pub fn new();
}

impl UhyveVm<KvmCpu> {
	pub fn run();
}
```

So whenever `UhyveVm::<T>::run()` is called, `T` is coerced to `KvmCpu` (or whatever the concrete type is), since on that platform no other implementor of `VirtualCPU` exists.

Consistency could be achieved either
- by making the signature of `UhyveVm::<T>::run()` independent of `KvmCpu`, forcing the user to explicitly supply `T` every time
- or by providing the type hint for `UhyveVm`.

I strongly support the latter, since I see the added verbosity as pure boilerplate code.

Regarding benchmarks: Changes in benchmark performance should not be expected to be reflected in public API types. Changes in performance can happen through any changes in the library. This should be first and foremost the responsibility of the changelog.

Reply:
For me, this still boils down to a personal preference. I completely see why you favor this, but I, personally, don't like to be implicit on such an important design aspect here.

> Consistency could be achieved either
>
> -   by making the signature of UhyveVm::<T>::run() independent of KvmCpu, forcing the user to explicitly supply T every time

I thought about this before, and this would be a nice approach. The only problem here is `run` with a debugger attached, which is on first sight not that trivial to make it part of the `VirtualCPU` trait.

I'm all open to improving that part, but I'd suggest doing that, once other debugger interfaces (e.g., whatever MacOS uses) get implemented.


> Regarding benchmarks: Changes in benchmark performance should not be expected to be reflected in public API types. Changes in performance can happen through any changes in the library. This should be first and foremost the responsibility of the changelog.

I totally agree when talking about smaller changes in performance due to implementation changes. But a change in the virtualization backend just because the default has changed can easily change the performance by orders of magnitude. This is not just an implementation detail for a library user, but a relevant choice when using `UhyveVM`.

Reply:
`Box<T, A = GlobalAlloc>`, `HashMap<K, V, S = RandomState>` and `LazyCell<T, F = fn() -> T>` all provide sensible defaults to avoid repeating yourself. Unnecessary `cfg's` should really be avoided. Something like having a cross-platform `AccelVcpu` and `EmulatedVcpu` or similar in the future and a default would at least get rid of that.

I really cannot see how this would be an improvement for anyone. I can only see this making everyone's life more difficult (maintainers and users of the library). While I am strongly against this, the behavior and current use should at the very least be consistent and not change on whether `rustc` can infer `T` through `run`.

Reply:
This is resolved now, right?